From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <lynxplay101@gmail.com>
Date: Tue, 4 Oct 2022 19:20:06 +0200
Subject: [PATCH] Expand de/serializeEntity API for adventure-nbt

With the addition of adventure-nbt, an API level representation of nbt
compounds are available to plugins.
While they are not widely used, entity storage is a perfect place for
them.

This commit adds a new de- and serialization method for entities that
returns the adventure-nbt representation of the compound tag.

diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 0bad57f184515734c49ee3030d167bd63d10ee0c..7db8ddd87405f306739fe5d938f564f8a3cc4d1f 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -273,5 +273,44 @@ public interface UnsafeValues {
     // Paper end
     // KTP start
     String itemStackAsString(@org.jetbrains.annotations.NotNull ItemStack itemStack);
-    // KTP stop
+
+    /**
+     * Serializes the entity passed into an adventure compound binary tag.
+     *
+     * @param entity the entity to serialize.
+     *
+     * @return the compound binary tag containing all required information to reconstruct the entity later on.
+     */
+    default net.kyori.adventure.nbt.CompoundBinaryTag serializeEntityToTag(final org.bukkit.entity.Entity entity) {
+        try (
+            final var inputStream = new java.io.ByteArrayInputStream(serializeEntity(entity));
+        ) {
+            return net.kyori.adventure.nbt.BinaryTagIO.reader().read(inputStream, net.kyori.adventure.nbt.BinaryTagIO.Compression.GZIP);
+        } catch (final java.io.IOException exception) {
+            throw new java.lang.IllegalArgumentException("The passed entity could not be serialized", exception);
+        }
+    }
+
+    /**
+     * Deserializes an entity given an adventure compound binary tag.
+     *
+     * @param data         the data tag that contains all information required to reconstruct the entity.
+     * @param world        the world into which the entity should be placed.
+     * @param preserveUUID a simple flag indicating if the uuid of the entity should be preserved.
+     *
+     * @return the deserialized entity instance.
+     */
+    default org.bukkit.entity.Entity deserializeEntity(final net.kyori.adventure.nbt.CompoundBinaryTag data,
+                                                       final World world,
+                                                       final boolean preserveUUID) {
+        try (
+            final var outputStream = new java.io.ByteArrayOutputStream();
+        ) {
+            net.kyori.adventure.nbt.BinaryTagIO.writer().write(data, outputStream, net.kyori.adventure.nbt.BinaryTagIO.Compression.GZIP);
+            return this.deserializeEntity(outputStream.toByteArray(), world, preserveUUID);
+        } catch (final java.io.IOException exception) {
+            throw new java.lang.IllegalArgumentException("The passed entity could not be deserialized", exception);
+        }
+    }
+    // KTP end
 }
